spec/google/gax/operation_spec.rb

```diff
@@ -0,0 +1,249 @@
+      called = false
+      get_method = proc do |_, options|
+        called = true
+        expect(options).to be_a(Google::Gax::CallOptions)
+        expect(options.retry_options.backoff_settings).to eq(backoff_settings)
+        GrpcOp.new(done: true, response: ANY)
+      end
+      mock_client = MockLroClient.new(get_method: get_method)
+      op = create_op(
+        GrpcOp.new(done: false, name: 'name'), client: mock_client
+      )
+      expect(called).to eq(false)
+      op.reload!(backoff_settings: backoff_settings)
+      expect(called).to eq(true)
+    end
+
+    it 'should yield the registered callback after the operation completes' do
+      op = create_op(GrpcOp.new(done: false), client: DONE_ON_GET_CLIENT)
+      called = false
+      op.on_done do |operation|
+        expect(operation.results).to eq(ANY)
+        called = true
+      end
+      expect(called).to eq(false)
+      expect(op.done?).to eq(false)
+      op.reload!
+      expect(called).to eq(true)
+      expect(op.done?).to eq(true)
+    end
+
+    it 'should yield the registered callbacks in order they were called.' do
+      op = create_op(GrpcOp.new(done: false), client: DONE_ON_GET_CLIENT)
+      expected_order = [1, 2, 3]
+      called_order = []
+      expected_order.each do |i|
+        op.on_done do |operation|
+          expect(operation.results).to eq(ANY)
+          called_order.push(i)
+        end
+      end
+      expect(called_order).to eq([])
+      expect(op.done?).to eq(false)
+      op.reload!
+      expect(called_order).to eq(expected_order)
+      expect(op.done?).to eq(true)
+    end
+  end
+
+  context 'method `wait_until_done!`' do
+    it 'should retry until the operation is done' do
+      to_call = 3
+      get_method = proc do
+        to_call -= 1
+        done = to_call == 0
+        GrpcOp.new(done: done, response: ANY)
+      end
+      mock_client = MockLroClient.new(get_method: get_method)
+      op = create_op(GrpcOp.new(done: false), client: mock_client)
+      op.wait_until_done!
+      expect(to_call).to eq(0)
+    end
+
+    it 'should wait until the operation is done' do
+      to_call = 3
+      get_method = proc do
+        to_call -= 1
+        done = to_call == 0
+        GrpcOp.new(done: done, response: ANY)
+      end
+      mock_client = MockLroClient.new(get_method: get_method)
+      op = create_op(GrpcOp.new(done: false), client: mock_client)
+      op.wait_until_done!
+      expect(to_call).to eq(0)
+    end
+
+    it 'times out' do
+      backoff_settings = BackoffSettings.new(0, 0, 0, 0, 0, 0, 1)
+      get_method = proc { GrpcOp.new(done: false) }
+      mock_client = MockLroClient.new(get_method: get_method)
+      op = create_op(GrpcOp.new(done: false), client: mock_client)
+      begin
+        op.wait_until_done!(backoff_settings: backoff_settings)
+      rescue Google::Gax::RetryError => exc
+        expect(exc).to be_a(Google::Gax::RetryError)
+      end
+    end
+
+    # TODO: Test exponential backoff.
```

Comment:
Include a link to an issue number for fixing this, unless you plan to do it in this PR